collector.go

```diff
@@ -191,25 +201,47 @@ func pduToSample(indexOids []int, pdu *gosnmp.SnmpPDU, metric *config.Metric, oi
 		// It's some form of string.
 		t = prometheus.GaugeValue
 		value = 1.0
-		stringType = true
-	}
-
-	labelnames := make([]string, 0, len(labels)+1)
-	labelvalues := make([]string, 0, len(labels)+1)
-	for k, v := range labels {
-		labelnames = append(labelnames, k)
-		labelvalues = append(labelvalues, v)
-	}
-	// For strings we put the value as a label with the same name as the metric.
-	// If the name is already an index, we do not need to set it again.
-	if stringType {
+		if len(metric.RegexpExtracts) > 0 {
+			v, ok := pdu.Value.(string)
+			if !ok {
+				log.Errorf("Invalid PDU value type: got %v, want string", reflect.TypeOf(pdu.Value))
+				return nil
+			}
+			return applyRegexExtracts(metric, v, labelnames, labelvalues)
+		}
+		// For strings we put the value as a label with the same name as the metric.
+		// If the name is already an index, we do not need to set it again.
 		if _, ok := labels[metric.Name]; !ok {
 			labelnames = append(labelnames, metric.Name)
 			labelvalues = append(labelvalues, pduValueAsString(pdu, metric.Type))
 		}
 	}
-	return prometheus.MustNewConstMetric(prometheus.NewDesc(metric.Name, metric.Help, labelnames, nil),
-		t, value, labelvalues...)
+
+	return []prometheus.Metric{prometheus.MustNewConstMetric(prometheus.NewDesc(metric.Name, metric.Help, labelnames, nil),
+		t, value, labelvalues...)}
+}
+
+func applyRegexExtracts(metric *config.Metric, pduValue string, labelnames, labelvalues []string) []prometheus.Metric {
+	results := []prometheus.Metric{}
+	for name, strMetricSlice := range metric.RegexpExtracts {
+		for _, strMetric := range strMetricSlice {
+			indexes := strMetric.Regex.FindStringSubmatchIndex(pduValue)
+			if indexes == nil {
+				continue
+			}
+			res := strMetric.Regex.ExpandString([]byte{}, strMetric.Value, pduValue, indexes)
+			v, err := strconv.ParseFloat(string(res), 64)
+			if err != nil {
+				log.Debugf("Error parsing float64 from value: %v", res)
```

Comment:
Including the metric name here would be useful, plus other log lines where you can